decompiler/pipeline/controlflowanalysis/variable_name_generation.py

```diff
@@ -107,62 +60,84 @@ class HungarianScheme(RenamingScheme):
         Integer: {8: "ch", 16: "s", 32: "i", 64: "l", 128: "i128"},
     }
 
-    custom_var_names = {"tmp_": "Tmp", "loop_break": "LoopBreak"}
-
     def __init__(self, task: DecompilerTask) -> None:
-        super().__init__(task)
-        self._name = VariableNameGeneration.name
-        self._var_name: str = task.options.getstring(f"{self._name}.variable_name", fallback="Var")
-        self._pointer_base: bool = task.options.getboolean(f"{self._name}.pointer_base", fallback=True)
-        self._type_separator: str = task.options.getstring(f"{self._name}.type_separator", fallback="")
-        self._counter_separator: str = task.options.getstring(f"{self._name}.counter_separator", fallback="")
-
-    def renameVariableNames(self):
-        """Rename all collected variables to the hungarian notation."""
-        for var in self._variables:
-            if self.alread_renamed(var._name):
-                continue
-            counter = _get_var_counter(var.name)
-            var._name = self._hungarian_notation(var, counter if counter else "")
-
-    def _hungarian_notation(self, var: Variable, counter: int) -> str:
-        """Return hungarian notation to a given variable."""
-        return f"{self._hungarian_prefix(var.type)}{self._type_separator}{self.custom_var_names.get(var._name.rstrip(counter), self._var_name)}{self._counter_separator}{counter}"
-
-    def _hungarian_prefix(self, var_type: Type) -> str:
+        self._task = task
+        self._var_name: str = task.options.getstring(f"{VariableNameGeneration.name}.variable_name", fallback="var")
+        self._pointer_base: bool = task.options.getboolean(f"{VariableNameGeneration.name}.pointer_base", fallback=True)
+        self._type_separator: str = task.options.getstring(f"{VariableNameGeneration.name}.type_separator", fallback="")
+        self._counter_separator: str = task.options.getstring(f"{VariableNameGeneration.name}.counter_separator", fallback="")
+
+        self._variables = self._get_variables_to_rename()
+
+        counter = Counter[tuple[str, str]]()
+        self._variable_rename_map: dict[VariableIdentifier, str] = {}
+
+        variable_id: VariableIdentifier
+        vars: list[Variable]
+        for variable_id, vars in self._variables.items():
+            # because the way our cfg works, each use site of each variable could theoretically have a different type
+            # we just take the first assuming that they are all the same...
```

Comment:
I am not sure about this. I thought out-of-SSA gives two variables only the same name, if they have the same type. The only exception could be aliased variables.

I am not happy with the first type, could we do something like the largest type or so? This way the order is important and I am afraid we get some non-deterministic behavior this way.

Reply:
Sure we can change it.
But what exactly do you mean by largest type? 
Most common in vars?

Reply:
Yes, most common was my idea, but I am open to suggestions

Reply:
I think that's the most reasonable criteria.
Done with next commit.